winch/codegen/src/isa/x64/masm.rs

```diff
@@ -1163,15 +1174,40 @@ impl MacroAssembler {
 
         Self {
             sp_offset: 0,
+            sp_max: 0,
+            stack_max_use_regions: SmallVec::new(),
             asm: Assembler::new(shared_flags.clone(), isa_flags.clone()),
             flags: isa_flags,
             shared_flags,
             ptr_size: ptr_type.into(),
         }
     }
 
+    /// Add the maximum stack used to a register, recording an obligation to update the
+    /// add-with-immediate instruction emitted to use the real stack max when the masm is being
+    /// finalized.
+    fn add_stack_max(&mut self, reg: Reg) {
+        let open = self.asm.buffer_mut().start_patchable();
+        // NOTE: we write out a value that is not encodable as an imm8, to ensure that we pick the
+        // variation of the x64 add instruction that supports imm32
+        self.asm.add_ir(-129, reg, OperandSize::S64);
+        let region = self.asm.buffer_mut().end_patchable(open);
+        self.stack_max_use_regions.push(region);
+    }
+
+    /// Overwrite a use of add-with-immediate in the machinst buffer, replacing the immediate with
+    /// the real stack size.
+    fn update_stack_max(&mut self, region: PatchRegion) {
+        let bytes = region.patch(self.asm.buffer_mut());
+        debug_assert_eq!(bytes.len(), 7);
+        debug_assert_eq!(bytes[0], 0x49);
+        debug_assert_eq!(bytes[1], 0x81);
+        bytes[3..].copy_from_slice(self.sp_max.to_le_bytes().as_slice());
+    }
+
     fn increment_sp(&mut self, bytes: u32) {
         self.sp_offset += bytes;
+        self.sp_max = self.sp_max.max(self.sp_offset);
```

Comment:
Could we add a comment here on why we're using `.max`? The reason why I bring this up is that we currently don't have a way to track the "canonical stack pointer offset" across the compiler, which is what we'd probably want to use here instead (that's what `.max` is doing here if I'm understanding correctly). We currently use the `sp_offset` as the single source of truth, but lately I've been thinking that it might be a good idea to change that (but not a blocker for this change!). This is mostly relevant in control flow, where we have to  soft-reset the stack pointer in some situations.

Reply:
The intent was to track the largest stack offset throughout code generation, so that we had an upper bound for the stack space needed. As push and reserve_stack both use `increment_sp`, this felt like a pretty good proxy for tracking the maximum seen stack allocation, but maybe I'm missing something here?

Reply:
> this felt like a pretty good proxy for tracking the maximum seen stack allocation

I think it is! I was just thinking that maybe it makes sense to have a comment stating that we do this (`.max`) given that we currently don't explicitly track the maximum amount of stack generated anywhere in the compiler yet, which I think we could as an optimization.

Reply:
Ah! I'll add a comment to clarify the intent here. I can also put some more documentation into the `sp_max` field itself 👍

Reply:
I added some more comments to this point in `increment_sp`, and to the definition of `sp_max`, would you mind taking another look?

Reply:
Looks great, thanks!